src/sparsegrad.jl

```diff
@@ -0,0 +1,69 @@
+###calculate jacobian by NiLang###
+"""
+    helper function params and fitparams
+        function params wraps parameters in track of gradient when given array/matrix.
+            params(v::DenseArray)
+            params(A::SparseMatrixCSC)
+        function fitparams fits the original datastructure with parameters in track of gradient.
+            fitparams(v::DenseArray)
+            fitparams(A::SparseMatrixCSC)
+"""
+params(v::DenseArray{<:Real}) = v
+fitparams(v::DenseArray{<:Real}, p) = p
+params(v::DenseArray{<:Complex{T}}) where T = collect(reinterpret(T, v))
+fitparams(v::DenseArray{<:T}, p::DenseArray{T2}) where {T<:Complex,T2} = collect(reinterpret(Complex{T2}, p))
+params(A::SparseMatrixCSC) = params(A.nzval)
+fitparams(A::SparseMatrixCSC, p) = SparseMatrixCSC(A.m, A.n, A.colptr, A.rowval, fitparams(A.nzval, p))
+
+"""
+    mvloss(l, j, y::AbstractArray{<:Real}, A, x) -> scalar l 
+mvloss defines loss on sparse-matrix & dense vector/matrix multiplication(Real Type)   
+"""
+@i function mvloss(l, j, y::AbstractArray{<:Real}, A, x)
+    imul!(y, A, x, 1.0, 1.0)
+    l += y[j]
+end
+
+"""
+    mvloss(l, j, y::AbstractArray{<:Complex}, A, x) -> scalar l 
+mvloss defines loss on sparse-matrix & dense vector/matrix multiplication(Complex Type)   
+"""
+@i function mvloss(l, j, y::AbstractArray{<:Complex}, A, x)
+    imul!(y, A, x, 1.0, 1.0)
+    if j%2 == 1
+        l += y[div(j-1,2)+1].re
+    else
+        l += y[div(j-1,2)+1].im
+    end
+end
+
+"""
+    nilang_mv_jacobian(A, x::AbstractArray) -> AbstractArray J
+nilang_mv_jacobian calculates the jacobian with respect to parameter A and x based on mvloss in NiLang
+"""
+function nilang_mv_jacobian(A, x::AbstractArray)
+    px, pA = params(x), params(A)
+    J = zeros(eltype(px), length(px), length(px) + length(pA))
+    for j=1:length(px)
+        _, _, _, _, gA, gx = NiLang.AD.Grad(mvloss)(Val(1), 0.0, j, zero(x), A, x)
+        J[j,:] = vcat(grad(params(gA)), grad(params(gx)))
+    end
+    return J
+end
+
+"""
+    forwarddiff_mv_jacobian(A, x) -> AbstractArray J 
+forwarddiff_mv_jacobian calculates the jacobian with respect to parameter A and x based on multiplication in ForwardDiff
+"""
+function forwarddiff_mv_jacobian(A, x)
```

Comment:
This is general multiplication, not just `mv`. This function can also be used to test matrix-matrix multiplication.

Reply:
Right, I'm going to add matrix- matrix multiplication in tests.  May be mm is a better name for the loss?(keep the same routine with pytorch? )

Reply:
If you want to change a name, please do it in the following style

```julia
julia> _f(x) = x
_f (generic function with 1 method)

julia> f(x::Vector) = _f(x)
f (generic function with 2 methods)

julia> g(x::Matrix) = _f(x)
g (generic function with 2 methods)
```

One of the golden rule of programming is DRY: don't repeat yourself, it means duplicated input in an interface, code copying are not recommended. If you change a name here, you have to copy the code.

Another comment is, in a test, you can write the code in a more casual style, for example, you do not need to document functions, since these functions will not be exported. Just add some comments to help understanding.
```julia
# calculates the jacobian of multiplication with ForwardDiff
# inputs can be matrices and vectors.
```

The above advices are for maintainability. When you change an implementation, you need to change
* its documentation
* every copy of this code

Most people fail to keep the documentations up to date if the documentation is too detailed, gradually, the software can be not usable. Unless you have a team maintaining your project and clients paying for your software, do not write unnecessary documentation. Note, the above comment remains valid even if your function name and arguments list are changed, unlike a docstring.

Reply:
Thanks for the delicate guidance! 
Besides the documentation,  I was intended to rename the loss function since in some multiplication cases the type of result array is SparseCSC and we need to implement corresponding loss function in another way (but now I feel such case could be distinguished by type).
``` julia
function mmloss(l, j, y::AbstractArray{<:Complex}, A, x)
function mmloss(l, j, y::AbstractSparseMatrix{<:Complex}, A, x)
```
And nilang_mm_jacobian is a general pipeline in calculating jacobian, so would it be better to let the loss be the input argument so as to adapt to more loss cases rather implement the whole jacobian function?

```julia
function nilang_mm_jacobian(A, x::AbstractArray, loss)
    px, pA = params(x), params(A)
    J = zeros(eltype(px), length(px), length(px) + length(pA))
    for j=1:length(px)
        _, _, _, _, gA, gx = NiLang.AD.Grad(loss)(Val(1), 0.0, j, zero(x), A, x)
        J[j,:] = vcat(grad(params(gA)), grad(params(gx)))
    end
    return J
end
```

Reply:
Weeks ago I complained about the current Nilang AD API for normal users, maybe it's time to revisit https://github.com/GiggleLiu/NiLang.jl/blob/f24036c478ace467ae71f630e1db2698e38fd64d/src/autodiff/jacobian.jl#L27-L44?

@GiggleLiu maybe open an issue for this gradient/jacobian API?

Reply:
For this PR, since it has the tendency to be epic, I feel it's perhaps okay to just leave `nilang_mm_jacobian` as it is so that this PR doesn't last forever.

Reply:
That's right. We shouldn't implement another jacobian method in this repo, just use the API that NiLang provides.